### PR TITLE
Use type_adv_tree in some filters to support addons adding advancements to existing unit types

### PR DESCRIPTION
**Pull request by newfrenchy83**

### CODE REVIEW

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg [hunk 1]
```diff
@@ -130,8 +130,9 @@
                                 description= _ "This massive blade was created centuries ago by long-forgotten elvish forgemasters, who imbued the bluish steel with an inner magical fire. Tongues of flame dance on the surface, giving the metal a flawless mirrored finish."
                                 cannot_use_message= _ "Only the leader of an army can wield this sword!"
                                 [filter]
-                                    type=Fighter,Commander,Lord,Princess,Battle Princess,Elvish Captain,Elvish Hero,Elvish Marshal,Elvish Champion,Paladin,Elvish Lord,Elvish High Lord
+                                    type_adv_tree=Fighter,Princess,Elvish Captain,Elvish Hero,Paladin,Elvish Lord
                                     x,y={X},{Y}
+                                    side=1
```
Comment (ProditorMagnus) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg:135]: ok to add but doesnt do anything

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg [hunk 2]
```diff
@@ -217,7 +218,7 @@
                                 [filter]
                                     side=1
                                     x,y={X},{Y}
-                                    type=Fighter,Commander,Lord,Princess,Battle Princess,Elvish Lord,Elvish High Lord,Elvish Fighter,Elvish Captain,Elvish Hero,Elvish Marshal,Elvish Champion,Elvish Ranger,Elvish Avenger,Horseman,Knight,Lancer,Paladin,Grand Knight,Dwarvish Fighter,Dwarvish Steelclad,Dwarvish Lord,Dwarvish Guardsman,Dwarvish Stalwart,Dwarvish Sentinel
+                                    type_adv_tree=Fighter,Princess,Elvish Fighter,Paladin,Elvish Lord,Elvish Ranger,Horseman,Dwarvish Fighter,Dwarvish Guardsman
```
Comment (ProditorMagnus) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg:221]: wasnt Paladin advancement of Horseman

File: data/campaigns/Liberty/scenarios/07_Glory.cfg [hunk 3]
```diff
@@ -348,7 +348,7 @@ One score and eight years ago, in 70 YW, the King of Wesnoth and the Lord of Ele
         name=turn 8
         [role]
             role=farseer
-            type=Outlaw,Footpad,Bandit,Thug,Thief,Huntsman,Trapper,Poacher,Rogue Mage,Rogue,Shadow Mage
+            type_adv_tree=Outlaw,Footpad,Bandit,Thug,Thief,Huntsman,Trapper,Poacher,Rogue Mage,Rogue,Shadow Mage
```
Comment (ProditorMagnus) [inline data/campaigns/Liberty/scenarios/07_Glory.cfg:351]: contains duplicates

File: data/campaigns/Son_Of_The_Black_Eye/scenarios/04_The_Siege_of_Barag_Gor.cfg [hunk 4]
```diff
@@ -218,7 +218,7 @@
         [/recall]
 
         [role]
-            type=Direwolf Rider,Goblin Knight,Goblin Pillager,Wolf Rider
+            type_adv_tree=Direwolf Rider,Goblin Knight,Goblin Pillager,Wolf Rider
```
Comment (ProditorMagnus) [inline data/campaigns/Son_Of_The_Black_Eye/scenarios/04_The_Siege_of_Barag_Gor.cfg:221]: duplicates

File: data/campaigns/Son_Of_The_Black_Eye/scenarios/11_Clash_of_Armies.cfg [hunk 5]
```diff
@@ -215,7 +215,7 @@
         [/fire_event]
 
         [role]
-            type=Orcish Warlord,Troll Warrior,Orcish Warrior,Troll,Orcish Grunt
+            type_adv_tree=Orcish Warlord,Troll Warrior,Orcish Warrior,Troll,Orcish Grunt
```
Comment (ProditorMagnus) [inline data/campaigns/Son_Of_The_Black_Eye/scenarios/11_Clash_of_Armies.cfg:218]: duplicates

File: data/campaigns/Son_Of_The_Black_Eye/scenarios/11_Clash_of_Armies.cfg [hunk 6]
```diff
@@ -224,7 +224,7 @@
 
         [role]
             role=cannonfodder2
-            type=Orcish Crossbowman,Orcish Slayer,Orcish Archer,Orcish Assassin
+            type_adv_tree=Orcish Crossbowman,Orcish Slayer,Orcish Archer,Orcish Assassin
```
Comment (ProditorMagnus) [inline data/campaigns/Son_Of_The_Black_Eye/scenarios/11_Clash_of_Armies.cfg:227]: duplicates

File: data/campaigns/Son_Of_The_Black_Eye/scenarios/14_Back_Home.cfg [hunk 7]
```diff
@@ -252,7 +252,7 @@
         [/recall]
 
         [role]
-            type=Orcish Warlord,Orcish Warrior,Orcish Grunt,Direwolf Rider,Goblin Pillager,Goblin Knight,Wolf Rider
+            type_adv_tree=Orcish Warlord,Orcish Warrior,Orcish Grunt,Direwolf Rider,Goblin Pillager,Goblin Knight,Wolf Rider
```
Comment (ProditorMagnus) [inline data/campaigns/Son_Of_The_Black_Eye/scenarios/14_Back_Home.cfg:255]: duplicates

File: data/campaigns/Son_Of_The_Black_Eye/scenarios/18_Northern_Alliance.cfg [hunk 8]
```diff
@@ -228,7 +228,7 @@
         [/recall]
 
         [recall]
-            type=Orcish Warlord,Orcish Slurbow,Troll Warrior,Direwolf Rider,Goblin Pillager,Orcish Crossbowman,Orcish Warrior,Goblin Knight
+            type_adv_tree=Orcish Warlord,Orcish Slurbow,Troll Warrior,Direwolf Rider,Goblin Pillager,Orcish Crossbowman,Orcish Warrior,Goblin Knight
```
Comment (ProditorMagnus) [inline data/campaigns/Son_Of_The_Black_Eye/scenarios/18_Northern_Alliance.cfg:231]: duplicates

**Comment by sevu** [comment]
When used inside [role] or other places such as [message] where exactly one is needed:
I think it uses the first one matching in `type`. As higher levels are listed first, they will be chosen.
With `type_adv_tree`, the lowest levels seem to be first now, so I suspect that now low level units are preferred as advisor,

Need to be tested if this is really the case though.

Good things on the other hand is that it supports mods better.

**Comment by ProditorMagnus** [comment]
Sounds like it would be useful to separate filtering and ordering.

Change all `type` to new name like `type_priority` and remove special handling of `type`, and make `type_priority` default to `type`

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg (hunk 1 above)
Comment (CelticMinstrel) [reply data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg:135]: Might be to satisfy the schema validator?

Comment (newfrenchy83) [reply data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: removed

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg (hunk 2 above)
Comment (newfrenchy83) [reply data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: fixed

File: data/campaigns/Son_Of_The_Black_Eye/scenarios/04_The_Siege_of_Barag_Gor.cfg (hunk 4 above)
Comment (newfrenchy83) [reply data/campaigns/Son_Of_The_Black_Eye/scenarios/04_The_Siege_of_Barag_Gor.cfg]: removed all[role] modifications

**Review by ProditorMagnus — APPROVED** [review]

File: data/campaigns/Descent_Into_Darkness/scenarios/03_A_Haunting_in_Winter.cfg [hunk 9]
```diff
@@ -1380,7 +1380,7 @@ All sense of time and being are entombed within darkness. How long has passed si
 
         [modify_unit]
             [filter]
-                type=Wolf,Great Wolf,Direwolf
+                type_adv_tree=Wolf
```
Comment (soliton-) [inline data/campaigns/Descent_Into_Darkness/scenarios/03_A_Haunting_in_Winter.cfg:1383]: Wolf unit types do not advance this way unless modified to do so. (Which they are not in DiD.)

Comment (newfrenchy83) [reply data/campaigns/Descent_Into_Darkness/scenarios/03_A_Haunting_in_Winter.cfg:1383]: unless the modification takes place in a modification type extension like Reign of the Lord which offers unit_types in the extension of the mainline lines and this is the reason for this request, to allow the event to take place when a unit advances beyond the mainline without forcing addon developers to look for specific solutions limited to non-general unit_types

Comment (newfrenchy83) [reply data/campaigns/Descent_Into_Darkness/scenarios/03_A_Haunting_in_Winter.cfg:1383]: excuse me, i hacen't not see what Wolf advance is null, i will fix that immediately.

Comment (soliton-) [reply data/campaigns/Descent_Into_Darkness/scenarios/03_A_Haunting_in_Winter.cfg]: The WML still needs to work for vanilla mainline...

You can use `type_adv_tree` if you really think that is useful (do you expect a modification to add more enemy guardian units of Wolf advancements?) but it needs to include all three wolf unit types.

Comment (newfrenchy83) [reply data/campaigns/Descent_Into_Darkness/scenarios/03_A_Haunting_in_Winter.cfg]: Done

**Pull request merged**